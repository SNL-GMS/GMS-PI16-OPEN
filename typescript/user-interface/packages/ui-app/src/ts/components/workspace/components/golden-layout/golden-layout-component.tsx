import { Intent, NonIdealState, Spinner } from '@blueprintjs/core';
import { VERSION_INFO } from '@gms/common-util';
import { AppActions, CommonWorkspaceOperations, CommonWorkspaceTypes } from '@gms/ui-state';
import React from 'react';
import { useDispatch } from 'react-redux';

import { authenticator } from '~app/authentication';
import { Mutations, Queries } from '~components/client-interface';
import { CommandPaletteContext } from '~components/common-ui/components/command-palette/command-palette-context';
import { Command, CommandScope } from '~components/common-ui/components/command-palette/types';

import { GoldenLayoutPanel } from './golden-layout-panel';
import { GoldenLayoutComponentProps } from './types';

export const GoldenLayoutComponent: React.FunctionComponent<GoldenLayoutComponentProps> = ({
  logo,
  openLayoutName,
  setOpenLayoutName,
  userName,
  setAuthStatus
}: GoldenLayoutComponentProps) => {
  const context = React.useContext(CommandPaletteContext);

  const setLayout = Mutations.SetLayoutMutation.useSetLayoutMutation();
  const userProfileQuery = Queries.UserProfileQuery.useUserProfileQuery();
  const dispatch = useDispatch();
  const registerCommands = React.useCallback(
    (commandsToRegister: Command[], scope: CommandScope) =>
      context?.registerCommands && context.registerCommands(commandsToRegister, scope),
    [context]
  );

  const logout = React.useCallback(() => {
    dispatch(AppActions.resetAppState());
    authenticator.logout(setAuthStatus);
  }, [dispatch, setAuthStatus]);

  const setGlDisplayState = React.useCallback(
    (displayName: string, displayState: CommonWorkspaceTypes.GLDisplayState) => {
      dispatch(CommonWorkspaceOperations.setGlDisplayState(displayName, displayState));
    },
    [dispatch]
  );

  const maybeUserProfile = userProfileQuery?.data ?? undefined;
  return maybeUserProfile ? (
    <GoldenLayoutPanel
      logo={logo}
      openLayoutName={openLayoutName}
      setOpenLayoutName={setOpenLayoutName}
      setGlDisplayState={setGlDisplayState}
      userName={userName}
      userProfile={maybeUserProfile}
      versionInfo={VERSION_INFO}
      setLayout={setLayout}
      registerCommands={registerCommands}
      logout={logout}
    />
  ) : (
    <NonIdealState
      action={<Spinner intent={Intent.PRIMARY} />}
      title="Loading Default Layout"
      description="Retrieving default layout for user..."
    />
  );
};
