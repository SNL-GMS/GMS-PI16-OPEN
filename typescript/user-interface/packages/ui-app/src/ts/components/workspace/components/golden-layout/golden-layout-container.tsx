import { compose } from '@gms/common-util';
import {
  AnalystWorkspaceActions,
  AppState,
  CommonWorkspaceActions,
  UserSessionOperations
} from '@gms/ui-state';
import * as ReactRedux from 'react-redux';
import { bindActionCreators } from 'redux';

import { GoldenLayoutComponent } from './golden-layout-component';
import { GoldenLayoutComponentProps } from './types';

// Map parts of redux state into this component as props
const mapStateToProps = (state: AppState): Partial<GoldenLayoutComponentProps> => ({
  currentTimeInterval: state.analystWorkspaceState.workflowState
    ? state.analystWorkspaceState.workflowState.timeRange
    : undefined,
  analysisMode: state.analystWorkspaceState.workflowState
    ? state.analystWorkspaceState.workflowState.analysisMode
    : undefined,
  openLayoutName: state.analystWorkspaceState.openLayoutName,
  keyPressActionQueue: state.commonWorkspaceState.keyPressActionQueue
});

// Map actions dispatch callbacks into this component as props
const mapDispatchToProps = (dispatch): Partial<GoldenLayoutComponentProps> =>
  bindActionCreators(
    {
      setOpenLayoutName: AnalystWorkspaceActions.setOpenLayoutName,
      setKeyPressActionQueue: CommonWorkspaceActions.setKeyPressActionQueue,
      setAuthStatus: UserSessionOperations.setAuthStatus
    } as any,
    dispatch
  );

/**
 * Connects the AppToolbar to the Redux store
 */
export const GoldenLayoutContainer = compose(
  ReactRedux.connect(mapStateToProps, mapDispatchToProps)
)(GoldenLayoutComponent);
