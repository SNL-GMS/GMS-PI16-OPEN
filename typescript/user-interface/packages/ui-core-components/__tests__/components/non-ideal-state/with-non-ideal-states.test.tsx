/* eslint-disable react/jsx-props-no-spreading */

import { Card, H5 } from '@blueprintjs/core';
import { Container } from '@gms/golden-layout';
import * as Enzyme from 'enzyme';
import * as React from 'react';

import {
  NonIdealStateDefinition,
  nonIdealStateWithNoSpinner,
  nonIdealStateWithSpinner
} from '../../..';
import {
  // NonIdealStateDefinition,
  WithNonIdealStates
} from '../../../src/ts/components/non-ideal-state';

const testNonIdealStateDefinitions: NonIdealStateDefinition<{
  status: any;
}>[] = [
  {
    condition: props => !props.status,
    element: nonIdealStateWithSpinner('No Data', 'Test')
  },
  {
    condition: props => props.status && props.status.loading,
    element: nonIdealStateWithSpinner('Loading:', 'Test')
  },
  {
    condition: props => props.status.group === undefined,
    element: nonIdealStateWithSpinner('No Group Data:', 'Test')
  },
  {
    condition: props =>
      props.status.group.Groups.length === 0 || props.status.group.status.length === 0,
    element: nonIdealStateWithNoSpinner('No Data:', 'Configure Groups')
  }
];

/// /////////////

const glWidth = 1000;
const glHeight = 500;

const myGLContainer: Container = {
  // Container
  width: glWidth,
  height: glHeight,
  parent: undefined,
  tab: undefined,
  title: 'container-title',
  layoutManager: undefined,
  isHidden: false,
  setState: jest.fn(),
  extendState: jest.fn(),
  getState: jest.fn(),
  getElement: jest.fn(),
  hide: jest.fn(),
  show: jest.fn(),
  setSize: jest.fn(),
  setTitle: jest.fn(),
  close: jest.fn(),
  // EventEmitter
  on: jest.fn(),
  emit: jest.fn(),
  trigger: jest.fn(),
  unbind: jest.fn(),
  off: jest.fn()
};

interface TestProps {
  glContainer: Container;
  status: {
    loading: boolean;
    error: any;
  };
}

interface TestState {
  status: any;
}

class TestComponent extends React.PureComponent<TestProps, TestState> {
  public render(): JSX.Element {
    return (
      <Card interactive>
        <H5>Card heading</H5>
        <p>Card content</p>
      </Card>
    );
  }
}

/// //////////

const CardNonIdealState = WithNonIdealStates(
  [...testNonIdealStateDefinitions] as any[],
  TestComponent
);

const myProps: TestProps = {
  glContainer: myGLContainer,
  status: {
    loading: true,
    error: undefined
  }
};

const wrapper = Enzyme.mount(
  // eslint-disable-next-line react/jsx-props-no-spreading
  <CardNonIdealState {...myProps} />
);

const myProps2: TestProps = {
  glContainer: myGLContainer,
  status: {
    loading: undefined,
    error: undefined
  }
};
const wrapper2 = Enzyme.mount(
  // eslint-disable-next-line react/jsx-props-no-spreading
  <CardNonIdealState {...myProps2} />
);

describe('With Non Ideal States', () => {
  it('to be defined', () => {
    expect(WithNonIdealStates).toBeDefined();
  });

  it('shallow renders', () => {
    const shallow = Enzyme.shallow(<TestComponent {...myProps} />);
    expect(shallow).toMatchSnapshot();
  });

  it('renders', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('bad props shallow renders', () => {
    const shallow2 = Enzyme.shallow(<TestComponent {...myProps2} />);
    expect(shallow2).toMatchSnapshot();
  });

  it('bad props renders', () => {
    expect(wrapper2).toMatchSnapshot();
    const spinner = wrapper2.find('Spinner');
    expect(spinner).toBeTruthy();
  });
});
