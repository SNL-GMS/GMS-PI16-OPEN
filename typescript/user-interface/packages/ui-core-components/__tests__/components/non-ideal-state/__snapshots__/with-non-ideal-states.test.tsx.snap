// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Non Ideal States bad props renders 1`] = `
<WrappedWithNonIdealStateChecker
  glContainer={
    Object {
      "close": [MockFunction],
      "emit": [MockFunction],
      "extendState": [MockFunction],
      "getElement": [MockFunction],
      "getState": [MockFunction],
      "height": 500,
      "hide": [MockFunction],
      "isHidden": false,
      "layoutManager": undefined,
      "off": [MockFunction],
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "show",
            [Function],
          ],
          Array [
            "resize",
            [Function],
          ],
          Array [
            "show",
            [Function],
          ],
          Array [
            "resize",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "parent": undefined,
      "setSize": [MockFunction],
      "setState": [MockFunction],
      "setTitle": [MockFunction],
      "show": [MockFunction],
      "tab": undefined,
      "title": "container-title",
      "trigger": [MockFunction],
      "unbind": [MockFunction],
      "width": 1000,
    }
  }
  status={
    Object {
      "error": undefined,
      "loading": undefined,
    }
  }
>
  <Blueprint3.NonIdealState
    action={
      <Blueprint3.Spinner
        intent="primary"
      />
    }
    description="Test"
    title="No Group Data:"
  >
    <div
      className="bp3-non-ideal-state"
    >
      <Component>
        <h4
          className="bp3-heading"
        >
          No Group Data:
        </h4>
      </Component>
      <div>
        Test
      </div>
      <Blueprint3.Spinner
        intent="primary"
      >
        <div
          className="bp3-spinner bp3-intent-primary"
        >
          <div
            className="bp3-spinner-animation"
          >
            <svg
              height={50}
              strokeWidth="8.00"
              viewBox="1.00 1.00 98.00 98.00"
              width={50}
            >
              <path
                className="bp3-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                className="bp3-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength={280}
                strokeDasharray="280 280"
                strokeDashoffset={210}
              />
            </svg>
          </div>
        </div>
      </Blueprint3.Spinner>
    </div>
  </Blueprint3.NonIdealState>
</WrappedWithNonIdealStateChecker>
`;

exports[`With Non Ideal States bad props shallow renders 1`] = `
<Blueprint3.Card
  elevation={0}
  interactive={true}
>
  <Component>
    Card heading
  </Component>
  <p>
    Card content
  </p>
</Blueprint3.Card>
`;

exports[`With Non Ideal States renders 1`] = `
<WrappedWithNonIdealStateChecker
  glContainer={
    Object {
      "close": [MockFunction],
      "emit": [MockFunction],
      "extendState": [MockFunction],
      "getElement": [MockFunction],
      "getState": [MockFunction],
      "height": 500,
      "hide": [MockFunction],
      "isHidden": false,
      "layoutManager": undefined,
      "off": [MockFunction],
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "show",
            [Function],
          ],
          Array [
            "resize",
            [Function],
          ],
          Array [
            "show",
            [Function],
          ],
          Array [
            "resize",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "parent": undefined,
      "setSize": [MockFunction],
      "setState": [MockFunction],
      "setTitle": [MockFunction],
      "show": [MockFunction],
      "tab": undefined,
      "title": "container-title",
      "trigger": [MockFunction],
      "unbind": [MockFunction],
      "width": 1000,
    }
  }
  status={
    Object {
      "error": undefined,
      "loading": true,
    }
  }
>
  <Blueprint3.NonIdealState
    action={
      <Blueprint3.Spinner
        intent="primary"
      />
    }
    description="Test"
    title="Loading:"
  >
    <div
      className="bp3-non-ideal-state"
    >
      <Component>
        <h4
          className="bp3-heading"
        >
          Loading:
        </h4>
      </Component>
      <div>
        Test
      </div>
      <Blueprint3.Spinner
        intent="primary"
      >
        <div
          className="bp3-spinner bp3-intent-primary"
        >
          <div
            className="bp3-spinner-animation"
          >
            <svg
              height={50}
              strokeWidth="8.00"
              viewBox="1.00 1.00 98.00 98.00"
              width={50}
            >
              <path
                className="bp3-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                className="bp3-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength={280}
                strokeDasharray="280 280"
                strokeDashoffset={210}
              />
            </svg>
          </div>
        </div>
      </Blueprint3.Spinner>
    </div>
  </Blueprint3.NonIdealState>
</WrappedWithNonIdealStateChecker>
`;

exports[`With Non Ideal States shallow renders 1`] = `
<Blueprint3.Card
  elevation={0}
  interactive={true}
>
  <Component>
    Card heading
  </Component>
  <p>
    Card content
  </p>
</Blueprint3.Card>
`;
